src/object.d

```diff
@@ -1431,6 +1431,132 @@ class TypeInfo_Struct : TypeInfo
     assert(!typeid(S).equals(&s, &s));
 }
 
+class RTTypeid(T) : TypeInfo
+if (is(T == struct))
+{
+    override string toString() const { return T.stringof; }
+
+    override bool opEquals(Object rhs)
+    {
+        if (this is rhs)
+            return true;
+        auto s = cast(const RTTypeid!T) rhs;
+        return s !is null;
+    }
+
+    override size_t getHash(scope const void* p) @trusted pure nothrow const
+    {
+        assert(p);
+        // Canonical prototype
+        alias HashFun = extern(D) size_t delegate() const nothrow @safe;
+        enum bool hasHash = is(typeof(&(cast(S*) null).toHash) == HashFun);
+        static if (hasHash)
+            return (cast(T*) p).toHash();
+        else
+            return hashOf(p[0 .. T.sizeof]);
+    }
+
+    override bool equals(in void* p1, in void* p2) @trusted pure nothrow const
```

Comment:
These qualifiers are inconsistent with `TypeInfo` and might collide with less qualified `opEquals` implementations.

```suggestion
    override bool equals(in void* p1, in void* p2)
```

Ditto below